container/shim/src/index.js

```diff
@@ -42,161 +47,231 @@ const PROXY_RESPONSE_HEADERS = [
 
-  app.get('/favicon.ico', (req, res) => {
-    res.sendStatus(404)
-  })
+app.get('/favicon.ico', (req, res) => {
+  res.sendStatus(404)
+})
+
+// Whenever nginx doesn't have a CAR file in cache, this is called
+app.get('/ipns/:cid', handleCID)
+app.get('/ipns/:cid/:path*', handleCID)
+app.get('/ipfs/:cid', handleCID)
+app.get('/ipfs/:cid/:path*', handleCID)
 
-  // Whenever nginx doesn't have a CAR file in cache, this is called
-  app.get('/ipns/:cid', handleCID)
-  app.get('/ipns/:cid/:path*', handleCID)
-  app.get('/ipfs/:cid', handleCID)
-  app.get('/ipfs/:cid/:path*', handleCID)
+async function handleCID (req, res) {
+  const cid = req.params.cid
+  const format = getResponseFormat(req)
 
-  async function handleCID (req, res) {
-    const cid = req.params.cid
-    const format = getResponseFormat(req)
+  debug(`Cache miss for ${req.path}`)
 
-    debug(`Cache miss for ${req.path}`)
+  res.set({
+    'Content-Type': mimeTypes.lookup(req.path) || 'application/octet-stream',
+    'Cache-Control': 'public, max-age=31536000, immutable',
+    'Saturn-Node-Id': nodeId,
+    'Saturn-Node-Version': NODE_VERSION
+  })
+
+  if (req.headers.range) {
+    let [start, end] = req.headers.range.split('=')[1].split('-')
+    start = parseInt(start, 10)
+    end = parseInt(end, 10)
 
     res.set({
-      'Content-Type': mimeTypes.lookup(req.path) || 'application/octet-stream',
-      'Cache-Control': 'public, max-age=31536000, immutable',
-      'Saturn-Node-Id': nodeId,
-      'Saturn-Node-Version': NODE_VERSION
+      'Accept-Ranges': 'bytes',
+      'Content-Range': `bytes ${start}-${end}/${testCAR.length}`
     })
+    return res.status(206).end(testCAR.slice(start, end + 1))
+  }
 
-    if (req.headers.range) {
-      let [start, end] = req.headers.range.split('=')[1].split('-')
-      start = parseInt(start, 10)
-      end = parseInt(end, 10)
+  // Testing CID
+  if (cid === TESTING_CID) {
+    return res.send(testCAR)
+  }
 
-      res.set({
-        'Accept-Ranges': 'bytes',
-        'Content-Range': `bytes ${start}-${end}/${testCAR.length}`
-      })
-      return res.status(206).end(testCAR.slice(start, end + 1))
-    }
+  debug(`Fetch ${req.path} from L2s`)
+  const cidHash = crypto.createHash('sha512').update(cid).digest()
+  Array.from(connectedL2Nodes.values())
+    .map(l2Node => ({
+      ...l2Node,
+      distance: xorDistance(
+        cidHash,
+        l2Node.idHash
+      )
+    }))
+    .sort((a, b) => xorDistance.compare(a.distance, b.distance))
+    .slice(0, 3)
+    .forEach(({ res }) => {
+      const payload = {
+        requestId: req.get('saturn-transfer-id'),
+        cid
+      }
+      res.write(`${JSON.stringify(payload)}\n`)
+    })
+
+  const onResponse = pDefer()
+  openCARRequests.set(cid, onResponse)
 
-    // Testing CID
-    if (cid === TESTING_CID) {
-      return res.send(testCAR)
+  let carResponse
+  try {
+    carResponse = await pTimeout(onResponse.promise, {
+      milliseconds: 10_000
+    })
+  } catch {}
```

Comment:
Should we print a debug log when an error is caught?

Reply:
Not sure, it would print one for every cache miss 🤔 We already have a line for this below: `debug(`Fetch ${req.path} from IPFS`)`, and the only error that can be thrown here is a timeout.

Reply:
Fair enough 👍🏻